metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java

```diff
@@ -268,18 +261,24 @@ void publishLogDelta(
         }
     }
 
-    void scheduleEmit(
+    void maybeScheduleEmit(
         String reason,
-        MetadataImage image
+        MetadataImage image,
+        boolean isOffsetBatchAligned
     ) {
-        resetSnapshotCounters();
+        if (isOffsetBatchAligned) {
+            this.disabledReason.compareAndSet("metadata image is not batch aligned", null);
```

Comment:
I'm pretty sure this will disable snapshotting for the lifetime of the server. Notice that we only clear this atomic ref during beginShutdown. 

Basically, we want to skip the snapshot if disabledReason is set _or_ isOffsetBatchAligned is false

Reply:
This also makes me realize we should have a unit test to cover bugs like this. Can we add a test that passes isOffsetBatchAligned=false and later passes isOffsetBatchAligned=true to ensure we will produce a snapshot once the metadata is batch-aligned?